File: Modules/Common/Techs/SIE.Web.Tech/Routings/Commands/PublishCommand.cs
using SIE.Security;
using SIE.Tech.Routings;
using SIE.Web.Command;

namespace SIE.Web.Tech.Routings.Commands
{
    /// <summary>
    /// 工艺路线发布命令
    /// </summary>
    [JsCommand("SIE.Web.Tech.Routings.Commands.PublishCommand")]
    [AllowAnonymous]
    public class PublishCommand : ViewCommand
    {
        /// <summary>
        /// 工艺路线发布命令执行方法
        /// </summary>
        /// <param name="args">视图参数</param>
        /// <param name="scope">作用域</param>
        /// <returns>执行结果</returns>
        protected override object Excute(ViewArgs args, string scope)
        {
            var obj = args.Data.ToJsonObject<RoutingLayout>();
            var version = RT.Service.Resolve<RoutingController>().ReleaseRoutingVersion(obj.Id, obj.Layout);
            string name = (version.IsDefault == YesNo.Yes ? "*" : string.Empty) + version.Name + "(" + version.ReferenceQty + ")";
            int.TryParse(version.Name.Substring(1), out int number);
            return new
            {
                Id = version.Id,
                text = name,
                leaf = true,
                nodetype = "VersionNode",
                routingId = version.RoutingId,
                layoutId = version.LayoutId,
                state = version.State,
                isDefault = version.IsDefault,
                referenceTime = version.ReferenceQty,
                versionName = version.Name,
                versionNum = number
            };
        }
    }
}

File: Modules/Common/Techs/SIE.Web.Tech/Routings/Commands/PublishCommand.js
SIE.defineCommand('SIE.Web.Tech.Routings.Commands.PublishCommand', {
    extend: 'SIE.cmd.Command',
    meta: { text: "发布", group: "edit", tooltip:"发布工艺路线，发布后无法更改".t(), iconCls: "icon-PlaylistCheck icon-blue" },
    execute: function (view, source) {
        var version = view.CurRoutingVersion;
        if (version.data.Id === null && version.data.isCopy) {
            SIE.Msg.showInstantMessage('不允许发布，工艺路线版本未保存'.t());
            return;
        }
        var drawCtrl = Ext.getCmp(view.routingDrawControlId);
        var layout = drawCtrl.getXml();
        if (view.layout.designControl.designCanvas.validateDesignData() === false)
            return;
        var iptdata = {
            Id: version.data.Id,
            Layout: layout
        };
        //客制化界面命令注册签名（注册在context，没有后台请求）
        SIE.Signature.createCmdContext({ command: Ext.getClassName(this), commandName: "发布", Type: view.model });
        view.execute({
            command: Ext.getClassName(this),
            data: iptdata,
            success: function (res) {
                var routingVersion = res.Result;
                var routingNode = version.parentNode;
                view.CurRoutingVersion.set('state', 1);
                view.CurRoutingVersion.set('isDefault', routingVersion.isDefault);
                if (routingVersion.isDefault === 0)
                    view.layout.routingControl.setDefaultCommandState(true);
                else {
                    routingNode.removeChild(version);
                    if (!routingNode) {
                        var routingControl = view.layout.routingControl;
                        routingNode = routingControl.getRootNode().childNodes.selectMany(function (p) { return p.childNodes; }).first(function (p) { return p.data.Id === routingVersion.routingId; });
                    }
                    var node = routingNode.appendChild(routingVersion);
                    var treePanel = routingNode.getOwnerTree();
                    treePanel.getSelectionModel().select(node);
                    var treeview = treePanel.getView();
                    if (routingNode && routingNode.parentNode) {
                        routingNode.expand();
                        routingNode.parentNode.expand();
                    }
                    treeview.focusRow(node);
                    treeview.fireEvent('itemdblclick', treeview, node, treeview.getNodeByRecord(node), treeview.indexOfRow(node));
                }

                view.layout.designControl.setMainBlockCommandDisabled('SaveCommand', true);
                view.layout.designControl.setMainBlockCommandDisabled('PublishCommand', true);
                view.layout.designControl.setMainBlockCommandDisabled('LeftRightCommand', true);
                view.layout.designControl.designCanvas.setLock(true);
                SIE.Msg.showInstantMessage('发布成功'.t());
            }
        });
    }
});